p2panda-sync/src/protocols.rs

```diff
@@ -0,0 +1,496 @@
+// SPDX-License-Identifier: AGPL-3.0-or-later
+
+use std::sync::{Arc, RwLock, RwLockReadGuard, RwLockWriteGuard};
+
+use futures::{Sink, SinkExt, Stream, StreamExt};
+use p2panda_core::extensions::DefaultExtensions;
+use p2panda_core::{Body, Header, Operation, PublicKey};
+use p2panda_store::{LogStore, MemoryStore, OperationStore};
+use serde::{Deserialize, Serialize};
+use tracing::debug;
+
+use crate::traits::{SyncError, SyncProtocol};
+
+type LogId = String;
+type SeqNum = u64;
+pub type LogHeights = Vec<(PublicKey, SeqNum)>;
+
+#[allow(clippy::large_enum_variant)]
+#[derive(Debug, Clone, Deserialize, Serialize)]
+pub enum Message<E = DefaultExtensions> {
+    Have(LogHeights),
+    Operation(Header<E>, Option<Body>),
+    SyncDone,
+}
+
+#[cfg(test)]
+impl<E> Message<E>
+where
+    E: Serialize,
+{
+    pub fn to_bytes(&self) -> Vec<u8> {
+        let mut bytes = Vec::new();
+        ciborium::into_writer(&self, &mut bytes).expect("type can be serialized");
+        bytes
+    }
+}
+
+#[derive(Clone, Debug, Default)]
+pub struct LogHeightSyncProtocol {
+    pub sync_done_sent: bool,
+    pub sync_done_received: bool,
+    pub store: Arc<RwLock<MemoryStore<LogId, DefaultExtensions>>>,
+}
+
+impl LogHeightSyncProtocol {
+    pub fn read_store(&self) -> RwLockReadGuard<MemoryStore<LogId, DefaultExtensions>> {
+        self.store.read().expect("error getting read lock on store")
+    }
+
+    pub fn write_store(&self) -> RwLockWriteGuard<MemoryStore<LogId, DefaultExtensions>> {
+        self.store
+            .write()
+            .expect("error getting write lock on store")
+    }
+}
+
+impl SyncProtocol for LogHeightSyncProtocol {
+    type Topic = LogId;
+    type Message = Message;
+
+    async fn run(
+        mut self,
+        topic: Self::Topic,
+        mut sink: impl Sink<Self::Message, Error = SyncError> + Unpin,
```

Comment:
Would be cool to use `Stream` instead of `Sink` here as well

Reply:
I need to look into this more, I'm not sure of the correct approach to replace a `Sink` with a `Stream`. Could I raise an issue and take it in a follow-up PR when I've done a little more research?